src/pybind/mgr/volumes/volume.py

```diff
@@ -0,0 +1,547 @@
+        with self.lock:
+            conn = self.connections.get(fs_name, None)
+            if conn:
+                conn.put_fs()
+
+    def _del_fs_handle(self, fs_name):
+        conn = self.connections.pop(fs_name, None)
+        if conn:
+            conn.del_fs()
+    def del_fs_handle(self, fs_name):
+        with self.lock:
+            self._del_fs_handle(fs_name)
+
+class VolumeClient(object):
+    def __init__(self, mgr):
+        self.mgr = mgr
+        self.connection_pool = ConnectionPool(self.mgr.rados)
+        self.purge_queue = ThreadPoolPurgeQueueMixin(self, 4)
+
+    def gen_pool_names(self, volname):
+        """
+        return metadata and data pool name (from a filesystem/volume name) as a tuple
+        """
+        return "cephfs.{}.meta".format(volname), "cephfs.{}.data".format(volname)
+
+    def get_fs(self, fs_name):
+        fs_map = self.mgr.get('fs_map')
+        for fs in fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs
+        return None
+
+    def get_mds_names(self, fs_name):
+        fs = self.get_fs(fs_name)
+        if fs is None:
+            return []
+        return [mds['name'] for mds in fs['mdsmap']['info'].values()]
+
+    def volume_exists(self, volname):
+        return self.get_fs(volname) is not None
+
+    def volume_exception_to_retval(self, ve):
+        """
+        return a tuple representation from a volume exception
+        """
+        return ve.to_tuple()
+
+    def create_pool(self, pool_name, pg_num, pg_num_min=None, pg_autoscale_factor=None):
+        # create the given pool
+        command = {'prefix': 'osd pool create', 'pool': pool_name, 'pg_num': pg_num}
+        if pg_num_min is not None:
+            command['pg_num_min'] = pg_num_min
+        r, outb, outs = self.mgr.mon_command(command)
+        if r != 0:
+            return r, outb, outs
+
+        # set pg autoscale if needed
+        if pg_autoscale_factor is not None:
+            command = {'prefix': 'osd pool set', 'pool': pool_name, 'var': 'pg_autoscale_bias',
+                       'val': str(pg_autoscale_factor)}
+            r, outb, outs = self.mgr.mon_command(command)
+        return r, outb, outs
+
+    def remove_pool(self, pool_name):
+        command = {'prefix': 'osd pool rm', 'pool': pool_name, 'pool2': pool_name,
+                   'yes_i_really_really_mean_it': True}
+        return self.mgr.mon_command(command)
+
+    def create_filesystem(self, fs_name, metadata_pool, data_pool):
+        command = {'prefix': 'fs new', 'fs_name': fs_name, 'metadata': metadata_pool,
+                   'data': data_pool}
+        return self.mgr.mon_command(command)
+
+    def remove_filesystem(self, fs_name):
+        command = {'prefix': 'fs rm', 'fs_name': fs_name, 'yes_i_really_mean_it': True}
+        return self.mgr.mon_command(command)
+
+    def create_mds(self, fs_name):
+        spec = orchestrator.StatelessServiceSpec()
```

Comment:
If you want, you can add proper constructor parameters to `StatelessServiceSpec`.

Reply:
note that this is part of PR #28082 and that refactor just moved around volume related routines which were all in `modules.py`.

As per your earlier suggestion related to these routines closely mimicking the ones in dashboard, can we do that as part of another PR?